File: src/test/scala/com/exasol/cloudetl/scriptclasses/TableExportQueryGeneratorTest.scala
package com.exasol.cloudetl.scriptclasses

import java.nio.file.Files

import scala.jdk.CollectionConverters._

import com.exasol.cloudetl.storage.StorageProperties

import org.mockito.Mockito._

class TableExportQueryGeneratorTest extends PathTest {

  override def beforeEach(): Unit = {
    super.beforeEach()
    when(metadata.getNodeCount()).thenReturn(1)
    when(metadata.getMemoryLimit()).thenReturn(new java.math.BigInteger("2000000000")) // 2GB
    ()
  }

  test("generateSqlForExportSpec returns SQL statement") {
    val params = properties ++ Map("PARALLELISM" -> "iproc()")
    when(metadata.getScriptSchema()).thenReturn(schema)
    when(exportSpec.getParameters()).thenReturn(params.asJava)
    val srcCols = Seq("tbl.col_int", "c_bool", "c_char")
    when(exportSpec.getSourceColumnNames).thenReturn(srcCols.asJava)

    val storageProperties = StorageProperties(params)
    val bucketPath = storageProperties.getStoragePath()
    val stringPairs = storageProperties.mkString()

    val expectedSQLStatement =
      s"""SELECT
         |  $schema.EXPORT_TABLE(
         |    '$bucketPath', '$stringPairs', 'col_int.c_bool.c_char', col_int, c_bool, c_char
         |)
         |FROM
         |  DUAL
         |GROUP BY
         |  iproc();
         |""".stripMargin

    assert(TableExportQueryGenerator.generateSqlForExportSpec(metadata, exportSpec) === expectedSQLStatement)
    verify(metadata, atLeastOnce).getScriptSchema()
    verify(exportSpec, times(1)).getParameters()
    verify(exportSpec, times(1)).getSourceColumnNames()
  }

  test("generateSqlForExportSpec throws if required property is not set") {
    val newProperties = properties - "S3_ENDPOINT"
    when(metadata.getScriptSchema()).thenReturn(schema)
    when(exportSpec.getParameters()).thenReturn(newProperties.asJava)

    val thrown = intercept[IllegalArgumentException] {
      TableExportQueryGenerator.generateSqlForExportSpec(metadata, exportSpec)
    }
    assert(thrown.getMessage().startsWith("E-CSE-2"))
    assert(thrown.getMessage().contains("'S3_ENDPOINT' property value is missing."))
    verify(exportSpec, times(1)).getParameters()
    verify(exportSpec, never).getSourceColumnNames()
  }

  test("generateSqlForExportSpec throws if columns cannot be parsed (e.g, contains extra '.')") {
    when(metadata.getScriptSchema()).thenReturn(schema)
    when(exportSpec.getParameters()).thenReturn(properties.asJava)
    val srcCols = Seq("tbl.c_int.integer")
    when(exportSpec.getSourceColumnNames).thenReturn(srcCols.asJava)

    val thrown = intercept[TableExporterException] {
      TableExportQueryGenerator.generateSqlForExportSpec(metadata, exportSpec)
    }
    assert(thrown.getMessage().startsWith("E-CSE-17"))
    assert(thrown.getMessage().contains("from given column syntax 'tbl.c_int.integer'"))
    verify(metadata, atLeastOnce).getScriptSchema()
    verify(exportSpec, times(1)).getParameters()
    verify(exportSpec, times(1)).getSourceColumnNames()
  }

  private[this] def createDummyFiles(path: java.nio.file.Path): Seq[java.nio.file.Path] = {
    val files = Seq("a.parquet", "b.parquet", "c.parquet").map(path.resolve(_))
    files.foreach(Files.createFile(_))
    files
  }

  test("generateSqlForExportSpec keeps the path files if 'overwrite' parameter is not set") {
    val bucketPath = Files.createTempDirectory("bucketPath")
    val files = createDummyFiles(bucketPath)
    val newProperties = properties ++ Map(
      "BUCKET_PATH" -> s"file://${bucketPath.toUri.getRawPath}"
    )
    when(metadata.getScriptSchema()).thenReturn(schema)
    when(exportSpec.getParameters()).thenReturn(newProperties.asJava)
    TableExportQueryGenerator.generateSqlForExportSpec(metadata, exportSpec)
    assert(Files.exists(bucketPath) === true)
    assert(Files.list(bucketPath).findAny().isPresent() === true)
    assert(Files.list(bucketPath).count() === 3)

    files.foreach(Files.deleteIfExists(_))
    Files.delete(bucketPath)
  }

  test("generateSqlForExportSpec deletes the path files if 'overwrite' parameter is set") {
    val bucketPath = Files.createTempDirectory("bucketPath")
    createDummyFiles(bucketPath)
    val newProperties = properties ++ Map(
      "BUCKET_PATH" -> s"file://${bucketPath.toUri.getRawPath}",
      "OVERWRITE" -> "true"
    )
    when(metadata.getScriptSchema()).thenReturn(schema)
    when(exportSpec.getParameters()).thenReturn(newProperties.asJava)
    TableExportQueryGenerator.generateSqlForExportSpec(metadata, exportSpec)
    assert(Files.exists(bucketPath) === false)
  }

}

File: src/test/java/com/exasol/cloudetl/scriptclasses/TableExportQueryGeneratorTest.java
package com.exasol.cloudetl.scriptclasses;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.io.IOException;
import java.nio.file.Files;
import java.util.*;
import java.util.stream.Collectors;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.exasol.cloudetl.storage.StorageProperties;

class TableExportQueryGeneratorTest extends PathTest {
    @BeforeEach
    void beforeEach() {
        when(this.metadata.getNodeCount()).thenReturn(1L);
        when(this.metadata.getMemoryLimit()).thenReturn(new java.math.BigInteger("2000000000"));
    }

    @Test
    void generateSqlForExportSpecReturnsSqlStatement() {
        final Map<String, String> params = new HashMap<>(this.properties);
        params.put("PARALLELISM", "iproc()");
        when(this.metadata.getScriptSchema()).thenReturn(this.schema);
        when(this.exportSpec.getParameters()).thenReturn(params);
        when(this.exportSpec.getSourceColumnNames()).thenReturn(List.of("tbl.col_int", "c_bool", "c_char"));

        final StorageProperties storageProperties = new StorageProperties(params);
        final String bucketPath = storageProperties.getStoragePath();
        final String stringPairs = storageProperties.mkString();
        final String expectedSqlStatement = String.format("SELECT\n" +
"  %s.EXPORT_TABLE(\n" +
"    '%s', '%s', 'col_int.c_bool.c_char', col_int, c_bool, c_char\n" +
")\n" +
"FROM\n" +
"  DUAL\n" +
"GROUP BY\n" +
"  iproc();\n", this.schema, bucketPath, stringPairs);

        assertEquals(expectedSqlStatement, TableExportQueryGenerator.generateSqlForExportSpec(this.metadata, this.exportSpec));
        verify(this.metadata, atLeastOnce()).getScriptSchema();
        verify(this.exportSpec, times(1)).getParameters();
        verify(this.exportSpec, times(1)).getSourceColumnNames();
    }

    @Test
    void generateSqlForExportSpecThrowsIfRequiredPropertyIsNotSet() {
        final Map<String, String> newProperties = new HashMap<>(this.properties);
        newProperties.remove("S3_ENDPOINT");
        when(this.metadata.getScriptSchema()).thenReturn(this.schema);
        when(this.exportSpec.getParameters()).thenReturn(newProperties);

        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class,
                () -> TableExportQueryGenerator.generateSqlForExportSpec(this.metadata, this.exportSpec));
        assertTrue(thrown.getMessage().startsWith("E-CSE-2"));
        assertTrue(thrown.getMessage().contains("'S3_ENDPOINT' property value is missing."));
        verify(this.exportSpec, times(1)).getParameters();
        verify(this.exportSpec, never()).getSourceColumnNames();
    }

    @Test
    void generateSqlForExportSpecThrowsIfColumnsCannotBeParsed() {
        when(this.metadata.getScriptSchema()).thenReturn(this.schema);
        when(this.exportSpec.getParameters()).thenReturn(this.properties);
        when(this.exportSpec.getSourceColumnNames()).thenReturn(List.of("tbl.c_int.integer"));

        final TableExporterException thrown = assertThrows(TableExporterException.class,
                () -> TableExportQueryGenerator.generateSqlForExportSpec(this.metadata, this.exportSpec));
        assertTrue(thrown.getMessage().startsWith("E-CSE-34"));
        assertTrue(thrown.getMessage().contains("from given column syntax 'tbl.c_int.integer'"));
        verify(this.metadata, atLeastOnce()).getScriptSchema();
        verify(this.exportSpec, times(1)).getParameters();
        verify(this.exportSpec, times(1)).getSourceColumnNames();
    }

    @Test
    void generateSqlForExportSpecKeepsThePathFilesIfOverwriteParameterIsNotSet() throws IOException {
        final java.nio.file.Path bucketPath = Files.createTempDirectory("bucketPath");
        final List<java.nio.file.Path> files = createDummyFiles(bucketPath);
        final Map<String, String> newProperties = new HashMap<>(this.properties);
        newProperties.put("BUCKET_PATH", "file://" + bucketPath.toUri().getRawPath());
        when(this.metadata.getScriptSchema()).thenReturn(this.schema);
        when(this.exportSpec.getParameters()).thenReturn(newProperties);
        TableExportQueryGenerator.generateSqlForExportSpec(this.metadata, this.exportSpec);
        assertTrue(Files.exists(bucketPath));
        try (var list = Files.list(bucketPath)) {
            assertTrue(list.findAny().isPresent());
        }
        try (var list = Files.list(bucketPath)) {
            assertEquals(3, list.count());
        }

        for (final java.nio.file.Path file : files) {
            Files.deleteIfExists(file);
        }
        Files.delete(bucketPath);
    }

    @Test
    void generateSqlForExportSpecDeletesThePathFilesIfOverwriteParameterIsSet() throws IOException {
        final java.nio.file.Path bucketPath = Files.createTempDirectory("bucketPath");
        createDummyFiles(bucketPath);
        final Map<String, String> newProperties = new HashMap<>(this.properties);
        newProperties.put("BUCKET_PATH", "file://" + bucketPath.toUri().getRawPath());
        newProperties.put("OVERWRITE", "true");
        when(this.metadata.getScriptSchema()).thenReturn(this.schema);
        when(this.exportSpec.getParameters()).thenReturn(newProperties);
        TableExportQueryGenerator.generateSqlForExportSpec(this.metadata, this.exportSpec);
        assertFalse(Files.exists(bucketPath));
    }

    private List<java.nio.file.Path> createDummyFiles(final java.nio.file.Path path) throws IOException {
        final List<java.nio.file.Path> files = List.of("a.parquet", "b.parquet", "c.parquet").stream().map(path::resolve)
                .collect(Collectors.toList());
        for (final java.nio.file.Path file : files) {
            Files.createFile(file);
        }
        return files;
    }
}
